File: app/src/main/java/com/challenge/fidoreader/Util/MapList.java
package com.challenge.fidoreader.Util;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;

public class MapList<K, V> {
    ArrayList<K> key;
    HashMap<K, V> map;

    int expectedSize = 0;

    public MapList(){
        key = new ArrayList<>();
        map = new HashMap<>();
        expectedSize = 0;
    }

    public void add(K k, V v){
        key.add(k);
        map.put(k,v);
    }

    public V get(Object key){
        return map.get(key);
    }

    public Object getKey(int num){
        return key.get(num);
    }

    public Object getValue(int num){
        return map.get(key.get(num));
    }

    public void clear(){
        key.clear();
        map.clear();
    }

    public int getSize(){
        return key.size();
    }

    public int expectedSize(){
        return expectedSize;
    }

    public void setExpectedSize(int expectedSize){
        this.expectedSize = expectedSize;
    }



}

File: app/src/main/java/com/challenge/fidoreader/Util/MapList.kt
package com.challenge.fidoreader.Util

import java.util.*

class MapList<K, V> {
    private var key: ArrayList<K> ?= null
    private var map: HashMap<K, V>
    private var expectedSize:Int

    init {
        key = ArrayList()
        map = HashMap()
        expectedSize = 0
    }

    fun add(k: K, v: V) {
        key!!.add(k)
        map[k] = v
    }

    operator fun get(key: Any?): V? {
        return map[key]
    }

    fun getKey(num: Int): K {
        return key!!.get(num)
    }

    fun getValue(num: Int): Any? {
        return this.map[key!![num]]
    }

    fun clear() {
        key!!.clear()
        map.clear()
    }

    fun getSize(): Int{
        return key!!.size
    }

    fun expectedSize(): Int {
        return expectedSize
    }

    fun setExpectedSize(expectedSize: Int) {
        this.expectedSize = expectedSize
    }


}